Activity Review

brendanshean opened the pull request:
## WHAT
This branch adds a clustering analysis feature for Evidence research. It enables machine learning clustering of student responses using text embeddings to identify patterns and group similar answers together:
* New `Clustering` model (Evidence::Research::GenAI::Clustering) to store clustering analysis results
* New `EntryEmbedding` model to cache text embeddings for student responses
* React UI at /clustering with controls, history, and results visualization
* Integration with an external Python clustering service (FastAPI)
* Service classes for job orchestration, polling, and result processing
* CSV export, test set creation from clusters, and sampling utilities

## WHY
To support Evidence research workflows:
* Dataset Building - Sample representative examples from clusters to create balanced test datasets for LLM prompts
* Analysis at Scale - Handle large volumes of feedback history that would be impractical to review manually
* Iterative Research - Track clustering history, compare runs with different parameters, and refine analysis
* Pattern Discovery - Group similar responses to identify common misconceptions, answer patterns, or response types

## HOW
Architecture:
* Rails Backend orchestrates the workflow:
  - `ClusteringsController` handles CRUD + async job management
  - `ClusteringClient` communicates with the Python service
  - `ClusteringJobPoller` + `ClusteringJobProcessor` handle async polling
  - Results stored in clusterings table with JSON metadata
* Python Service (external app [Clustering](https://github.com/empirical-org/clustering))
  - Receives entry IDs, fetches embeddings, runs K-means/HDBSCAN
  - Optimizes cluster count via silhouette scores
  - Returns cluster assignments and statistics
* React Frontend (`ClusteringApp.tsx`):
  - Controls for algorithm, min/max clusters, sampling
  - History panel showing previous runs
  - Results table with cluster details and entry counts
  - Actions to export CSV or create test datasets

### Screenshots
(If applicable. Also, please censor any sensitive data)

### Notion Card Links
https://www.notion.so/quill/Clustering-Optimal-Suboptimal-Separately-Experiment-Clustering-Tool-UI-Updates-2c7d42e6f941801d83f2c0ad54e14e79?source=copy_link

### What have you done to QA this feature?
Extensive testing on staging running the clustering algorithm

PR Checklist | Your Answer
------------ | -------------
Have you added and/or updated tests? |  YES
Have you deployed to Staging? | YES
Self-Review: Have you done an initial self-review of the code below on Github? | YES

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+  const [nClusters, setNClusters] = React.useState(DEFAULT_N_CLUSTERS);
+  const [minClusters, setMinClusters] = React.useState(MIN_CLUSTERS);
+  const [maxClusters, setMaxClusters] = React.useState(DEFAULT_MAX_CLUSTERS);
+  const [responsesToCluster, setResponsesToCluster] = React.useState(maxResponses);
```
small nit, bit I'd add type definitions for the rest of these since you're already defining `ClusteringParams`

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+          <div className="radio-group algorithm-radio-group">
+            <RadioButton
+              label="K-Means (faster, requires specifying number of clusters)"
+              onClick={() => !isClustering && setAlgorithm(KMEANS)}
```
We try to avoid inline onclick functions when possible (inline functions create a new function instance on every render, so there a performance implications). I'm not sure if it'll be much of an issue for this particular page, but I wanted to flag that the standard convention would be more along the lines of this:

```
function handleSetAlgorithmToKmeans()
  !isClustering && setAlgorithm(KMEANS)
end

<RadioButton
  onClick={handleSetAlgorithmToKmeans}
 ...
```

I think there are several instances here where the logic is basically the same save for the value, but since you pass the value to `RadioButton` already, I think you can just pass this function for whichever RadioButtons concern setting the algorithm:

```
function handleSetAlgorithm(e)
  !isClustering && setAlgorithm(e.target.value)
end
```

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+              <div className="radio-group algorithm-radio-group">
+                <RadioButton
+                  label="Auto-optimize (find best number of clusters)"
+                  onClick={() => !isClustering && algorithm === KMEANS && setMode(AUTO)}
```
see above comment:

```
function handleSetMode(e)
  !isClustering && setAlgorithm(e.target.value)
end
```

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+                <div className="radio-option-with-input">
+                  <RadioButton
+                    label="Fixed number of clusters"
+                    onClick={() => !isClustering && algorithm === KMEANS && setMode(FIXED)}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+                      disabled={isClustering}
+                      max={AUTO_MAX_CLUSTERS}
+                      min={MIN_CLUSTERS}
+                      onChange={(e) => setNClusters(parseInt(e.target.value, 10))}
```
see above comment, same applies for onChange handlers:

```
function handleSetNClusters(e)
  setNClusters(parseInt(e.target.value, 10))}
end
```

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+                <div className="radio-option-with-input">
+                  <RadioButton
+                    label="Custom range"
+                    onClick={() => !isClustering && algorithm === KMEANS && setMode(RANGE)}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+                        disabled={isClustering}
+                        max={ABSOLUTE_MAX_CLUSTERS - 1}
+                        min={MIN_CLUSTERS}
+                        onChange={(e) => handleMinChange(parseInt(e.target.value, 10))}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+                        disabled={isClustering}
+                        max={ABSOLUTE_MAX_CLUSTERS}
+                        min={MIN_CLUSTERS + 1}
+                        onChange={(e) => handleMaxChange(parseInt(e.target.value, 10))}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+            </div>
+            <RadioButton
+              label="HDBSCAN (density-based, automatically finds clusters)"
+              onClick={() => !isClustering && setAlgorithm(HDBSCAN)}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+            disabled={isClustering}
+            max={maxResponses}
+            min={1}
+            onChange={(e) => setResponsesToCluster(Math.min(parseInt(e.target.value, 10) || 1, maxResponses))}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/history/clusteringHistory.tsx:
```diff
@@ -0,0 +1,158 @@
+      }
+      if (filters.algorithm !== ALL) {
+        const algorithm = clustering.parameters?.algorithm || KMEANS;
+        if (algorithm !== filters.algorithm) {
```
if possible, I think we should hoist the `algorithm` const up so that we can just do:

```
if (filters.algorithm !== ALL && algorithm !== filters.algorithm) {
  return false;
}
```

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/history/clusteringHistory.tsx:
```diff
@@ -0,0 +1,158 @@
+      <div className="history-header-row">
+        <button
+          className="expand-toggle"
+          onClick={() => setIsExpanded(!isExpanded)}
```
see above comment about inline onclick functions

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/history/clusteringHistory.tsx:
```diff
@@ -0,0 +1,158 @@
+
+        <div className="history-filters">
+          <select
+            onChange={(e) => setFilters(prev => ({ ...prev, status: e.target.value }))}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/history/clusteringHistory.tsx:
```diff
@@ -0,0 +1,158 @@
+          </select>
+
+          <select
+            onChange={(e) => setFilters(prev => ({ ...prev, algorithm: e.target.value }))}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/history/clusteringHistory.tsx:
```diff
@@ -0,0 +1,158 @@
+          {hasActiveFilters && (
+            <button
+              className="clear-filters-btn"
+              onClick={() => setFilters({ status: ALL, algorithm: ALL })}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/history/clusteringHistory.tsx:
```diff
@@ -0,0 +1,158 @@
+                    key={clustering.id}
+                    onClick={() => {
+                      window.location.href = clusteringPageUrl(evidenceActivityId, conjunction, clustering.id);
+                    }}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/results/clusterTable.tsx:
```diff
@@ -0,0 +1,128 @@
+              <div className="cluster-header">
+                <div
+                  className="cluster-header-left"
+                  onClick={() => onToggleCluster(group.clusterId)}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/results/clusterTable.tsx:
```diff
@@ -0,0 +1,128 @@
+                    /* eslint-disable-next-line jsx-a11y/click-events-have-key-events, jsx-a11y/no-noninteractive-element-interactions */
+                    <label
+                      className="exemplars-checkbox"
+                      onClick={(e) => e.stopPropagation()}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/results/clusterTable.tsx:
```diff
@@ -0,0 +1,128 @@
+                    >
+                      <input
+                        checked={showExemplars}
+                        onChange={() => onToggleShowExemplars(group.clusterId)}
```
see above comment

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/results/clusterTable.tsx:
```diff
@@ -0,0 +1,128 @@
+        {clusterGroups.map((group) => {
+          const isExpanded = expandedClusters.has(group.clusterId);
+          const centroidResponse = group.exemplars.length > 0 ? group.exemplars[0] : group.responses[0];
+          const displayName = centroidResponse?.text || `Cluster ${group.clusterId}`;
```
we can destructure all of these values to avoid repetition:

`const { exemplars, responses, clusterId, etc } = group`

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/history/clusteringHistory.tsx:
```diff
@@ -0,0 +1,158 @@
+            <div className="history-cards">
+              {filteredHistory.map((clustering) => {
+                const isCurrentClustering = clustering.id === Number(currentClusteringId);
+                const isHdbscan = clustering.parameters?.algorithm === HDBSCAN;
```
we can destructure all of these values to avoid repetition:

`const { id, parameters, status, etc } = clustering`

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/hooks/useClusteringPolling.ts:
```diff
@@ -0,0 +1,74 @@
+          onRefetchHistory();
+        }
+      } catch {
+        // Continue polling on error
```
do we want to still maybe add some kind of messaging (maybe in the console) if this errors?

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/main/index.tsx:
```diff
@@ -0,0 +1,200 @@
+                <p><strong>Entry Embeddings:</strong> {activityData.clusteringEntryCount || 0}</p>
+              </div>
+              <button
+                className="quill-button-archived primary contained medium save-embeddings-btn"
```
use `quill-button` rather than archived

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/__tests__/__snapshots__/clusteringControls.test.tsx.snap:
```diff
@@ -0,0 +1,180 @@
+        <button
+          class="quill-button-archived primary contained large "
+          type="submit"
+        >
```
I think we should use `quill-button` here rather than archived since I think we will eventually remove that from the component library

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/__tests__/__snapshots__/resultsSection.test.tsx.snap:
```diff
@@ -0,0 +1,304 @@
+      <button
+        class="quill-button-archived secondary outlined small"
+        type="button"
+      >
```
use `quill-button` rather than archived

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/controls/clusteringControls.tsx:
```diff
@@ -0,0 +1,194 @@
+
+        <div className="submit-section">
+          <button
+            className={`quill-button-archived primary contained large ${isClustering ? 'disabled' : ''}`}
```
use `quill-button` rather than archived

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/components/results/resultsSection.tsx:
```diff
@@ -0,0 +1,260 @@
+            className="quill-button-archived secondary outlined small"
+            onClick={onReset}
+            type="button"
+          >
```
use `quill-button` rather than archived

eadams17 commented on services/QuillLMS/client/app/bundles/Staff/components/evidence/clustering/main/index.tsx:
```diff
@@ -0,0 +1,200 @@
+              >
+                {isSavingEmbeddings ? BUTTON_SAVING : isProcessingEmbeddings ? 'Processing...' : BUTTON_SAVE_EMBEDDINGS}
+              </button>
+              <p className="help-text" style={{ marginTop: '6px' }}>
```
we should add this style to the `.help-text` class in place of inline styling

eadams17 reviewed (changes requested):
Overall the frontend portion looks pretty solid! The main changes I think we should make are around inline onclick handlers that I go into more detail with in one of my comments. There are also a few instances where we should be using the new quill button since the archived one will eventually be phased out.

dandrabik commented on services/QuillLMS/engines/evidence/app/models/evidence/research/clustering_integration.md:
```diff
@@ -0,0 +1,125 @@
+# Clustering Service Integration
```
Nice Readme!

dandrabik commented on services/QuillLMS/engines/evidence/app/services/evidence/research/gen_ai/clustering_job_processor.rb:
```diff
@@ -0,0 +1,141 @@
+  module Research
+    module GenAi
+      class ClusteringJobProcessor < ApplicationService
+        SUBOPTIMAL_CLUSTER_ID_OFFSET = 1000
```
This might be worth revisiting to see if there is a way to store this without using an offset OR have some sort of loud error if the size of the optimal set is more than this number.

dandrabik commented on services/QuillLMS/engines/evidence/app/services/evidence/research/gen_ai/test_set_dataset_creator.rb:
```diff
@@ -0,0 +1,60 @@
+          sampled_entries.each do |sampled_entry|
+            feedback_history = sampled_entry[:clustering_entry].feedback_history
+
+            DraftExample.create!(
```
Noting that this is errorable operation in a loop, so this code is not idempotent. If this is called by a job and it errors mid-way through, it'll keep creating duplicates up to the errored record.

dandrabik reviewed (approved):
Overall, my scan of the backend looks good (doing a cursory scan since this is only staff facing and is mostly generated code). I left a few comments that are worth looking at, but nothing is blocking.

brendanshean committed:
Clustering

brendanshean committed:
Dataset Editor and Clustering Tool upgrades (#13603)

* Add source column for Exampls and hover over in Dataset Editor

* wip

* Move summary statistics

* Get rid of chevrons

* Move back to the right

* potential optimization

* Attempt to speed up ID page

* Clean up sessionId reference

* Cleanup so diffs

* Spec failures

* Cleanup old table

* Linter

* test errors

* Add feedback_session_uid and feedback histories dummy

brendanshean committed:
Update annotations

brendanshean committed:
Activity review QA (#13635)

* Update annotations

* Use relative numbering for clustering history labels

Instead of showing global database IDs (e.g. #10, #8, #7), clustering
history cards now display activity-relative numbers (#5, #4, #3) based
on chronological order within each activity.

* Center summary statistics box in clustering header

* Nest clustering routes under activity-review path

* Add sample ID to review dropdown labels

* Reverse review dropdown order so most recent is first

* Revert "Update annotations"

This reverts commit 3b4d26a1f96ab64610b63504d32a2fc458a738d3.

* Update Clustering snapshot for summary statistics markup change

* Fix misspelled keyword argument in add_more action

* Add missing index swap to copied buckets migration

* Refetch representative samples when URL references uncached sample

After creating a representative sample, the hash-based navigation
doesn't trigger a full page reload, so React Query serves stale
cached data (staleTime: Infinity). The dropdown showed the previous
sample as "Current" instead of the newly created one.

* Activity review QA part2 (#13644)

* Address Activity Review QA items

- Update heading to use standard comprehension-page-header-container pattern
- Rename title to "Activity Review - [LLM Optimal/Suboptimal/Import Preview]"
- Rearrange layout: remove max-width constraint, tighten radio button
  spacing, display scoring stats horizontally, full-width stem
- Add "Show only matched (green)" and "Show only mismatched (red)" filters
- Add snackbar notification on Add to Import Preview click
- Rename button from "Send all selected to import preview" to
  "Add to Import Preview"
- Preserve optimal/suboptimal colors when rows move to Import Preview
- Render feedback HTML instead of showing raw markup tags
- Fix ClusterSampler to use largest remainder method so proportional
  sampling reliably produces exactly 100 samples

* Update Activity Review colors for consistency and colorblind accessibility

- Active tab buttons: use $quill-green (#06806B) to match other action buttons
- Assignment green: use $quill-green-vibrant-50 (#79C933) for better contrast
- Assignment red: use $quill-red (#D3443D) for better contrast

* Persist review state when switching samples or navigating away

Flush pending debounced save immediately when switching between
representative samples or unmounting the component. Clear the
react-query cache for the old sample so returning fetches fresh
server data with the saved review state.

* Add dataset selection modal for Activity Review import

Let users choose which dataset to duplicate when importing from
Activity Review instead of always using the most recent one. The
selected dataset also drives the bucket dropdown so buckets
recalculate when the target changes.

* Update target dataset label to 'Importing to: vX.Y'

* Add Date Imported column, fix import notes/feedback, refresh buckets on dataset change

- Add importedToDatasets field and Date Imported column to all Activity Review tabs
- Rename Date column to Date Written
- Strip HTML tags from feedback when importing to dataset editor
- Fix review_state params to use to_unsafe_h so nested data (notes) persists through import
- Refresh bucket dropdown when switching target dataset, preserving matching values
- Add spec for RepresentativeSampleDatasetImporter

* Store bucket assignments per dataset, fix review state clearing, fix notes display

- Change bucket storage from single string to per-dataset map (buckets: Record<datasetId, title>)
  so switching between import target datasets preserves each dataset's bucket assignments
- Auto-fill bucket when switching datasets if the same title exists in the new dataset
- Prevent review state from being cleared by guarding all save paths against empty rowsById
- Set skipSaveRef when switching samples to prevent auto-save with stale data
- Update rowsByIdRef before redirect after import to prevent cleanup overwriting timestamps
- Add notes to ExampleSerializer so dataset editor displays imported notes
- Update importer to read buckets map (with legacy bucket fallback)
- Persist selectedImportDatasetId in review state across page reloads

* Add comprehensive tests for Activity Review and representative samples

Frontend: ActivityReview component (49 tests), DatasetSelectionModal (20 tests)
Backend: RepresentativeSample factory, model spec, controller spec,
builder/expander/dataset importer service specs, serializer spec,
and OrganicExamplesImporter spec (68 tests)

* Fix rubocop offenses in Activity Review service and spec files

* Sanitize feedback HTML and trigger autosave on dataset change

brendanshean committed:
Add missing feedback_session_uid column to test structure.sql

brendanshean committed:
Clear representative samples list cache on unmount

The list query uses staleTime: Infinity, so navigating away and
returning served stale cached data that could be missing newly
created samples. Clearing the list cache on unmount forces a
fresh server fetch on the next visit.

brendanshean committed:
Add spacing between scoring stats and tab buttons in activity review

brendanshean committed:
Fix Get 100 more to sample from unselected responses only

ClusterSampler was sampling from all entries then filtering out
already-selected ones, causing diminishing returns on each click.
Now passes exclude_ids and target_total_samples so sampling only
draws from the pool of responses not yet selected.

brendanshean committed:
Use black text on green assignment buttons for better contrast

brendanshean committed:
Fix rubocop Lint/NoReturnInBeginEndBlocks in ClusterSampler

brendanshean committed:
Adjust activity review table column widths for better space distribution

brendanshean committed:
Fix dataset editor bucket title wrapping instead of scrolling

Widen bucket column (160→240px) and narrow notes (240→160px).
Fix text wrapping by removing height:auto !important override
that fought with InlineEditor's auto-sizing, and use pre-wrap
with overflow:hidden so scrollHeight calculation works properly.

brendanshean committed:
Refetch stem vaults when opening dataset selection modal

brendanshean committed:
Show most recent datasets first in selection modal

brendanshean committed:
Use O#/S# naming convention for bucket optimal detection

Parse bucket titles with O# prefix as optimal and S# prefix as
suboptimal. Titles matching neither pattern set optimal to nil.

brendanshean committed:
Require full O#:/S#: pattern for bucket optimal detection

Tighten regex to require digit(s) followed by a colon, matching
the expected naming convention (e.g. O1: label, S2: label).

brendanshean committed:
Add source_example_id to Example (#13645)

* Add source_example_id to preserve original example dates across dataset versions

When examples are copied between dataset versions (via dataset editor save
or activity review import), their created_at dates were resetting to the
copy time. This adds a source_example_id foreign key that tracks the root
original example, and an original_created_at serializer attribute so the
UI displays the stable original generation date.

* Add thor task to backfill source_example_id for existing examples

Walks datasets in version order per stem_vault, matching examples across
versions by student_response (type-agnostic since type can change between
versions). Supports --dry-run for previewing.

Usage: bundle exec thor backfill_source_example_ids:run

* Extract shared example for STI Example specs and add draft_example_spec

Common base class tests (associations, validations, readonly attrs) are
now in a shared 'an example' context used by all three STI subclasses.
Adds missing draft_example_spec and source_example association coverage
for prompt and draft examples.

* Update dummy structure.sql

* Fix failing tests: dataset selection modal order and dummy schema

Account for reversed dataset rendering order in datasetSelectionModal
tests and add missing feedback_session_uid column to evidence engine
dummy structure.sql.
